vendor/k8s.io/kubernetes/pkg/controller/daemon/patch_nodeselector.go

```diff
@@ -0,0 +1,89 @@
+package daemon
+
+import (
+	"k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	appsinformers "k8s.io/client-go/informers/apps/v1beta1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	extensionsinformers "k8s.io/client-go/informers/extensions/v1beta1"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+func NewNodeSelectorAwareDaemonSetsController(openshiftDefaultNodeSelectorString, kubeDefaultNodeSelectorString string, namepaceInformer coreinformers.NamespaceInformer, daemonSetInformer extensionsinformers.DaemonSetInformer, historyInformer appsinformers.ControllerRevisionInformer, podInformer coreinformers.PodInformer, nodeInformer coreinformers.NodeInformer, kubeClient clientset.Interface) (*DaemonSetsController, error) {
+	controller, err := NewDaemonSetsController(daemonSetInformer, historyInformer, podInformer, nodeInformer, kubeClient)
+	if err != nil {
+		return controller, err
+	}
+	controller.namespaceLister = namepaceInformer.Lister()
+	controller.namespaceStoreSynced = namepaceInformer.Informer().HasSynced
+	controller.openshiftDefaultNodeSelectorString = openshiftDefaultNodeSelectorString
+	if len(controller.openshiftDefaultNodeSelectorString) > 0 {
+		controller.openshiftDefaultNodeSelector, err = labels.Parse(controller.openshiftDefaultNodeSelectorString)
+		if err != nil {
+			return nil, err
+		}
+	}
+	controller.kubeDefaultNodeSelectorString = kubeDefaultNodeSelectorString
+	if len(controller.kubeDefaultNodeSelectorString) > 0 {
+		controller.kubeDefaultNodeSelector, err = labels.Parse(controller.kubeDefaultNodeSelectorString)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return controller, nil
+}
+
+func (dsc *DaemonSetsController) namespaceNodeSelectorMatches(node *v1.Node, ds *extensions.DaemonSet) (bool, error) {
+	if dsc.namespaceLister == nil {
+		return true, nil
+	}
+
+	// this is racy (different listers) and we get to choose which way to fail.  This should requeue.
+	ns, err := dsc.namespaceLister.Get(ds.Namespace)
+	if apierrors.IsNotFound(err) {
+		return false, err
+	}
+	// if we had any error, default to the safe option of creating a pod for the node.
```

Comment:
I think the safe option is **not** to create the Pod rather than creating it somewhere it shouldn't be. You can always create it in the next sync but undoing seems worse.

Also you are ignoring that error so we won't know where it's broken.

Reply:
> I think the safe option is not to create the Pod rather than creating it somewhere it shouldn't be. You can always create it in the next sync but undoing seems worse.

Think of this as an optimization.  If the method just returned true every time we'd function correctly and be really slow.  What we want is say `false` as often as we can without ever doing it falsely.

Reply:
I tend to think about it as correctness and security issue (although kubelet being a last line of defense here stops it). The project default node selector is a security feature preventing pods from being scheduled to some nodes. If we schedule them there just because we have encountered an error, that seems wrong.

> What we want is say false as often as we can without ever doing it falsely.

That would result in creating pod to restricted nodes and deleting them when we stop getting errors.

Reply:
> I tend to think about it as correctness and security issue (although kubelet being a last line of defense here stops it).

I don't think that we should think of our controllers as agents of security.  The security feature is that kubelets reject pods that aren't allowed to run on them.  The optimization in the controller is to avoid creating pods that will never succeed.

The controller can never be that agent of security, since the point of action is the kubelet

Reply:
1. [According to Clayton we restrict /bind to nodeName](https://github.com/kubernetes/kubernetes/issues/42002#issuecomment-297595485) so we shouldn't let DS controller bind to nodes where it shouldn't.

2. Correctness - you shall not create pods for nodes which are not targeted by DS (after implications from node selector admission). And this PR is actually about preventing that.

Take the extreme case. You have 1000 node cluster and default project nodeSelector targeting 5 of them. Because of an error you just created a 1000 Pods instead of 5 and you are going to delete those 995 when the error goes away. (Multiply that by the number of DS in the namespace/cluster.) That doesn't seem right. It puts unwanted load on etcd, scheduler, ...

I see this PR as fixing correctness, not an optimization - that's likely why are opinions differ in this particular case.

Reply:
> Correctness - you shall not create pods for nodes which are not targeted by DS (after implications from node selector admission). And this PR is actually about preventing that.

That is the wrong bar of correctness in this case.  If you were authoring the upstream DS controller, maybe.  We're shimming in an optimization, and the DS controller's correctness today is "create all the pods I might need".  Changing that here would cause incompatible carry behavior.

We have to fail to `true`.

Reply:
We have tests that ensure the kubelet rejects pods that do not match it's selector. *That* is the security boundary. This filtering is an optimization.

Reply:
Since I believe upsteam behaviour is broken here I don't agree that fallback